File: node/src/controllers/dashboardController.js
const { Transaction, Category, Account } = require('../models');
const { Op, fn, col, literal, Sequelize } = require('sequelize');

exports.summary = async (req, res, next) => {
  try {
    const userId = req.userId;
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const endOfMonth = new Date(now.getFullYear(), now.getMonth() + 1, 0, 23, 59, 59);

    const accounts = await Account.findAll({ where: { userId } });
    const totalBalance = accounts.reduce((s, a) => s + parseFloat(a.balance || 0), 0);

    const monthIncome = await Transaction.sum('amount', {
      where: {
        userId,
        type: 'receita',
        transactionDate: { [Op.between]: [startOfMonth, endOfMonth] },
      },
    });

    const monthExpenses = await Transaction.sum('amount', {
      where: {
        userId,
        type: 'despesa',
        transactionDate: { [Op.between]: [startOfMonth, endOfMonth] },
      },
    });

    const netBalance = (monthIncome || 0) - (monthExpenses || 0);

    res.json({
      totalBalance: parseFloat(totalBalance.toFixed(2)),
      monthlyIncome: parseFloat((monthIncome || 0).toFixed(2)),
      monthlyExpenses: parseFloat((monthExpenses || 0).toFixed(2)),
      netBalance: parseFloat(netBalance.toFixed(2)),
    });
  } catch (err) {
    next(err);
  }
};

exports.monthlyEvolution = async (req, res, next) => {
  try {
    const userId = req.userId;
    const year = req.query.year || new Date().getFullYear();

    const transactions = await Transaction.findAll({
      attributes: [
        [fn('DATE_TRUNC', 'month', col('transaction_date')), 'month'],
        'type',
        [fn('SUM', col('amount')), 'total'],
      ],
      where: {
        userId,
        transaction_date: {
          [Op.between]: [`${year}-01-01`, `${year}-12-31`],
        },
      },
      group: [fn('DATE_TRUNC', 'month', col('transaction_date')), 'type'],
      order: [[fn('DATE_TRUNC', 'month', col('transaction_date')), 'ASC']],
      raw: true,
    });

    const months = [];
    for (let i = 0; i < 12; i++) {
      months.push({
        month: i + 1,
        income: 0,
        expenses: 0,
      });
    }

    for (const t of transactions) {
      const monthIndex = new Date(t.month).getMonth();
      if (t.type === 'receita') {
        months[monthIndex].income = parseFloat(t.total || 0);
      } else {
        months[monthIndex].expenses = parseFloat(t.total || 0);
      }
    }

    res.json(months);
  } catch (err) {
    next(err);
  }
};

exports.calendar = async (req, res, next) => {
  try {
    const userId = req.userId;
    const year = parseInt(req.query.year) || new Date().getFullYear();
    const month = req.query.month ? parseInt(req.query.month) : null;

    const where = {
      userId,
      type: 'despesa',
      transaction_date: month
        ? { [Op.between]: [`${year}-${String(month).padStart(2, '0')}-01`, `${year}-${String(month).padStart(2, '0')}-31`] }
        : { [Op.between]: [`${year}-01-01`, `${year}-12-31`] },
    };

    const days = await Transaction.findAll({
      attributes: [
        [fn('DATE_TRUNC', 'day', col('transaction_date')), 'date'],
        [fn('SUM', col('amount')), 'total'],
        [fn('COUNT', col('id')), 'count'],
      ],
      where,
      group: [fn('DATE_TRUNC', 'day', col('transaction_date'))],
      order: [[fn('DATE_TRUNC', 'day', col('transaction_date')), 'ASC']],
      raw: true,
    });

    const result = days.map(d => ({
      date: new Date(d.date).toISOString().split('T')[0],
      total: parseFloat(d.total || 0),
      count: parseInt(d.count || 0),
    }));

    res.json(result);
  } catch (err) {
    next(err);
  }
};

exports.categoryBreakdown = async (req, res, next) => {
  try {
    const userId = req.userId;
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const endOfMonth = new Date(now.getFullYear(), now.getMonth() + 1, 0, 23, 59, 59);

    const spending = await Transaction.findAll({
      attributes: [
        [col('Transaction.category_id'), 'categoryId'],
        [fn('SUM', col('Transaction.amount')), 'total'],
        [fn('COUNT', col('Transaction.id')), 'count'],
      ],
      where: {
        userId,
        type: 'despesa',
        transactionDate: { [Op.between]: [startOfMonth, endOfMonth] },
      },
      include: [
        { model: Category, as: 'category', attributes: ['id', 'name', 'icon', 'color', 'type'] },
      ],
      group: ['Transaction.category_id', 'category.id'],
      raw: true,
    });

    const categories = spending.map(s => ({
      id: s['category.id'],
      name: s['category.name'],
      icon: s['category.icon'],
      color: s['category.color'],
      type: 'despesa',
      total: parseFloat(s.total || 0),
      count: parseInt(s.count || 0),
    }));

    const totalSpent = categories.reduce((sum, c) => sum + c.total, 0);
    for (const c of categories) {
      c.percentage = totalSpent > 0 ? parseFloat(((c.total / totalSpent) * 100).toFixed(1)) : 0;
    }

    res.json({ categories, totalSpent: parseFloat(totalSpent.toFixed(2)) });
  } catch (err) {
    next(err);
  }
};

File: python/app/controllers/dashboard_controller.py
from datetime import datetime, timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from sqlalchemy import func, extract

from app.database import get_db
from app.models.transaction import Transaction
from app.models.category import Category
from app.middleware.auth import get_current_user_id

router = APIRouter(prefix="/api/dashboard", tags=["dashboard"])


@router.get("/summary")
def summary(
    period: str = Query("month", regex="^(week|month|year)$"),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    now = datetime.utcnow()
    if period == "week":
        start_date = now - timedelta(days=7)
    elif period == "year":
        start_date = now - timedelta(days=365)
    else:
        start_date = now - timedelta(days=30)

    receitas = (
        db.query(func.coalesce(func.sum(Transaction.amount), 0))
        .filter(
            Transaction.user_id == user_id,
            Transaction.type == "receita",
            Transaction.transaction_date >= start_date,
        )
        .scalar()
    )

    despesas = (
        db.query(func.coalesce(func.sum(Transaction.amount), 0))
        .filter(
            Transaction.user_id == user_id,
            Transaction.type == "despesa",
            Transaction.transaction_date >= start_date,
        )
        .scalar()
    )

    total_transactions = (
        db.query(func.count(Transaction.id))
        .filter(
            Transaction.user_id == user_id,
            Transaction.transaction_date >= start_date,
        )
        .scalar()
    )

    balance = receitas - despesas

    return {
        "receitas": round(receitas, 2),
        "despesas": round(despesas, 2),
        "balance": round(balance, 2),
        "total_transactions": total_transactions,
        "period": period,
    }


@router.get("/monthly-evolution")
def monthly_evolution(
    year: int = Query(None),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    if year is None:
        year = datetime.utcnow().year

    results = (
        db.query(
            extract("month", Transaction.transaction_date).label("month"),
            Transaction.type,
            func.sum(Transaction.amount).label("total"),
        )
        .filter(
            Transaction.user_id == user_id,
            extract("year", Transaction.transaction_date) == year,
        )
        .group_by("month", Transaction.type)
        .order_by("month")
        .all()
    )

    months = []
    for r in results:
        months.append({
            "month": int(r.month),
            "type": r.type,
            "total": round(r.total, 2),
        })

    return {"year": year, "months": months}


@router.get("/category-breakdown")
def category_breakdown(
    start_date: str = Query(None),
    end_date: str = Query(None),
    type: str = Query("despesa", regex="^(receita|despesa)$"),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    query = (
        db.query(
            Category.name,
            Category.icon,
            Category.color,
            func.coalesce(func.sum(Transaction.amount), 0).label("total"),
        )
        .join(Transaction, Transaction.category_id == Category.id, isouter=True)
        .filter(Category.user_id == user_id, Category.type == type)
    )

    if start_date:
        query = query.filter(Transaction.transaction_date >= datetime.fromisoformat(start_date))
    if end_date:
        query = query.filter(Transaction.transaction_date <= datetime.fromisoformat(end_date))

    results = query.group_by(Category.id, Category.name, Category.icon, Category.color).all()

    total = sum(r.total for r in results)

    categories = []
    for r in results:
        percentage = round((r.total / total * 100), 2) if total > 0 else 0
        categories.append({
            "name": r.name,
            "icon": r.icon,
            "color": r.color,
            "total": round(r.total, 2),
            "percentage": percentage,
        })

    return {
        "type": type,
        "total": round(total, 2),
        "categories": categories,
    }
